### PR TITLE
Implementation of HC process

### PR DESCRIPTION
**Pull request by TomFischer**
Missing things:
- [x] Documentation of parameters
- [x] ~~Documentation of benchmarks~~ Marc is working on this, however code review can start.
- [x] goswamy benchmark

For discussion: 
- [x] process name: `HC` or `ComponentTransport` or ...
- [x] search length for boundary conditions

**Update** Changing the default search length breaks the benchmark 'ogs-TES_zeolite_discharge_small'. So I removed the commit.

### CODE REVIEW

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+            // \todo the first argument has to be changed for non constant
+            // porosity model
+            auto const porosity =
+                _process_data.porous_media_properties.getPorosity(t, pos)
+                    .getValue(0.0, C_int_pt);
+
+            auto const retardation_factor =
+                _process_data.retardation_factor(t, pos)[0];
+
+            auto KCC = local_K.template block<num_nodes, num_nodes>(0, 0);
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: It is possible to move lines 183 to 189 about the matrix blocks before the start of the integration loop?

File: ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        //! \ogs_file_param_special{prj__processes__process__ComponentTransport__process_variables__pressure}
+        "pressure"});
+
+    auto const& porous_medium_configs =
+        //! \ogs_file_param{prj__processes__process__ComponentTransport__porous_medium}
+        config.getConfigSubtree("porous_medium");
+    PorousMediaProperties porous_media_properties{
+        createPorousMediaProperties(mesh, porous_medium_configs)};
+
+    //! \ogs_file_param{prj__processes__process__ComponentTransport__fluid}
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp]: Suggest to use class FluidProperties, which contains density, viscosity, specific heat capacity and thermal conductivity (optional). There is a creator for FluidProperties,  please see
https://github.com/ufz/ogs/blob/master/ProcessLib/LiquidFlow/CreateLiquidFlowMaterialProperties.cpp#L49

File: ProcessLib/ComponentTransport/CreatePorousMediaProperties.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#include "MaterialLib/PorousMedium/Porosity/createPorosityModel.h"
+#include "MaterialLib/PorousMedium/Storage/createStorageModel.h"
+
+#include "MeshLib/Mesh.h"
+
+namespace ProcessLib
+{
+namespace ComponentTransport
+{
+PorousMediaProperties createPorousMediaProperties(
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/CreatePorousMediaProperties.cpp]: :+1: . This class can be general one for all fluid flow processes. After this PR merged, I  may refractory the other flow processes by using this class to reduce the source code  duplication  about porous medium among process classes.

File: ProcessLib/ComponentTransport/ComponentTransportProcessData.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        delete;
+
+    //! Assignments are not needed.
+    void operator=(ComponentTransportProcessData const&) = delete;
+
+    //! Assignments are not needed.
+    void operator=(ComponentTransportProcessData&&) = delete;
+
+    PorousMediaProperties porous_media_properties;
+    std::unique_ptr<MaterialLib::Fluid::FluidProperty> viscosity_model;
```
Comment (wenqing) [inline ProcessLib/ComponentTransport/ComponentTransportProcessData.h]: As the comment given about, you can replace
```
  std::unique_ptr<MaterialLib::Fluid::FluidProperty> viscosity_model;
  std::unique_ptr<MaterialLib::Fluid::FluidProperty> fluid_density;
```
with
```
  const std::unique_ptr<MaterialLib::Fluid::FluidProperties>
    _fluid_properties;
```
Here is an example to use FluidProperties:
https://github.com/ufz/ogs/blob/master/ProcessLib/LiquidFlow/LiquidFlowMaterialProperties.cpp#L75

**Comment by TomFischer** [comment]
I compared the run times on my laptop with and without the last commit (c0bfb08). The previous version (without `FluidProperties`) takes circa 79 seconds while the new version takes ~ 81.6 seconds.

**Comment by wenqing** [comment]
@TomFischer If the fluid properties do not depend on density, the extra operation to access members of  individual fluid property class is to access an entry of the property array
```
const std::array<std::unique_ptr<FluidProperty>, FluidPropertyTypeNumber>
        _property_models;
```
class FluidProperties enables the temperature density dependent fluid property models, e.g. WaterViscosityIAPWS.

File: MaterialLib/Fluid/PropertyVariableType.h [hunk 5]
```diff
@@ -22,7 +22,8 @@ enum class PropertyVariableType
     T = 0,                   ///< temperature.
     p = 1,                   ///< pressure.
     rho = p,                 ///< density. For some models, rho substitutes p
-    number_of_variables = 2  ///< Number of property variables.
+    C = 2,
```
Comment (endJunction) [inline MaterialLib/Fluid/PropertyVariableType.h]: short comment

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+                _process_data.solute_dispersivity_transverse(t, pos)[0];
+            auto const& solute_dispersivity_longitudinal =
+                _process_data.solute_dispersivity_longitudinal(t, pos)[0];
+
+            // Use the fluid density model to compute the density
+            vars[static_cast<int>(
+                MaterialLib::Fluid::PropertyVariableType::C)] = C_int_pt;
+            vars[static_cast<int>(
+                MaterialLib::Fluid::PropertyVariableType::p)] = p_int_pt;
+            auto const density_water = _process_data.fluid_properties->getValue(
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: better to call it density_liquid, it's not always water...

File: ProcessLib/ComponentTransport/ComponentTransportFEM.h [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+            auto const& decay_rate = _process_data.decay_rate(t, pos)[0];
+            auto const& molecular_diffusion_coefficient =
+                _process_data.molecular_diffusion_coefficient(t, pos)[0];
+
+            // Use the viscosity model to compute the viscosity
+            auto const viscosity =
+                _process_data.fluid_properties->getValue(
+                    MaterialLib::Fluid::FluidPropertyType::Viscosity, vars);
+
+            GlobalDimMatrixType const perm_over_visc =
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/ComponentTransportFEM.h]: Better to call it K_over_mu as it is done below.

File: ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        specific_body_force.resize(b.size());
+        std::copy_n(b.data(), b.size(), specific_body_force.data());
+    }
+
+    ComponentTransportProcessData process_data{
+        std::move(porous_media_properties),
+//        std::move(viscosity_model),
+        fluid_reference_density,
+//        std::move(fluid_density),
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/CreateComponentTransportProcess.cpp]: drop the comments or explain why needed

File: ProcessLib/ComponentTransport/CreatePorousMediaProperties.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        // i.e., the isotropic case is handled at the moment)
+        auto const& permeability_conf =
+            //! \ogs_file_param{prj__processes__process__ComponentTransport__porous_medium__porous_medium__permeability}
+            porous_medium_config.getConfigSubtree("permeability");
+        intrinsic_permeability_models.emplace_back(
+            MaterialLib::PorousMedium::createPermeabilityModel(
+                permeability_conf));
+
+        // Parameter for the specific storage.
+        auto const& storage_conf =
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/CreatePorousMediaProperties.cpp]: typo: config ;)

File: ProcessLib/ComponentTransport/PorousMediaProperties.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+          _specific_storage_models(std::move(other._specific_storage_models)),
+          _material_ids(other._material_ids)
+    {
+    }
+
+    MaterialLib::PorousMedium::Porosity const& getPorosity(
+        double t, SpatialPosition const& pos) const;
+
+    Eigen::MatrixXd const& getIntrinsicPermeability(double t,
+                                              SpatialPosition const& pos) const;
```
Comment (endJunction) [inline ProcessLib/ComponentTransport/PorousMediaProperties.h]: maybe formatting... `git clang-format ufz/master` is good!

**Pull request merged**